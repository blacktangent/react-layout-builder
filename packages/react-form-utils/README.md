# react-form-utils

[![Build Status](https://travis-ci.org/blacktangent/react-layout-builder.svg?branch=master)](https://travis-ci.org/blacktangent/react-layout-builder)

`react-form-utils` provides helper functions to retrieve form props and form values

```
import {inputPropsLookup, inputValueLookup, formSerialize} from 'react-form-utils';
```

### `inputPropsLookup(fieldsProps, fieldName) `

Argument    | Type        | Example
:-----------| :-----------| :-----------
`fieldsProps`      | `object`      | `{name: {type: 'text'}, email: {type: 'email'}}`
`fieldName`   | `string`      | `name`, `email[0]`, `email[1]`

#### returns
`object`: props, e.g. `{type: 'text'}`

#### e.g.
```js
const fieldsProps = {
  name: { type: 'text' },
  age_group: {
    type: 'select',
    size: 4,
    options: []
  }
  address: {
    type: 'nested',
    fields: {
      street: {
        type: 'text'
      }
    }
  }  
};
const fieldName = 'address[0][street]';

const props = inputPropsLookup(fieldsProps, fieldName);
// <input name={fieldName} {...props} />
```

### `formSerialize(form)`

    formSerialize(form)  

Argument    | Type        | Example
:-----------| :-----------| :-----------
`form`      | `DOMNode`      | `form` element

#### returns
`object`. serialized, e.g. `{type: 'text'}`

#### e.g.
```js
// <form ref={f => this.form =f}><input name="message" value="hello"/></form>

const values = formSerialize(this.form);
// { message: 'hello' }

```

### `inputValueLookup(values, fieldName) `

Argument    | Type        | Example
:-----------| :-----------| :-----------
`values`      | `object`      | `{address: [ {zip: '55555'} ]}`
`fieldName`   | `string`      | `address[0][zip]`

#### returns
`any`. input value, e.g. `{type: 'text'}`

#### e.g.
```js
const values = {
  name: 'Ali',
  age_group: '31-40'
  address: {
    0: {
      street: '13 Calle 49a'
     }
};
const fieldName = 'address[0][street]';

const value = inputValueLookup(values, fieldName);

// <input name={fieldName} value={value} />
```
