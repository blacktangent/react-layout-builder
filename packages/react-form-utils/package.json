{
  "name": "react-form-utils",
  "version": "1.0.15",
  "description": "Serialize/Deserialize form input, and lookup",
  "main": "build/index.js",
  "author": "code@alvinsj.net",
  "repository": {
    "type": "git",
    "url": "https://github.com/blacktangent/react-layout-builder"
  },
  "keywords": [
    "reactjs",
    "form",
    "layout",
    "builder",
    "react-component"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/blacktangent/react-layout-builder/issues"
  },
  "homepage": "https://github.com/blacktangent/react-layout-builder",
  "dependencies": {
    "form-serialize": "alvinsj/form-serialize#export-fn",
    "object-assign": "^4.x.x"
  },
  "scripts": {
    "jest": "NODE_PATH=src:NODE_PATH jest",
    "test": "jest",
    "jest:watch": "NODE_PATH=src:NODE_PATH jest --watch",
    "compile": "babel -d build/ src/",
    "prepublish": "babel -d build/ src/"
  }
}
