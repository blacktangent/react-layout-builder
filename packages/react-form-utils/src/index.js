import assign from 'object.assign';
import serialize, { hash_serializer } from 'form-serialize';

export const inputPropsLookup = (inputProps, inputName) => {
  var props = inputProps[inputName];
  if (props) return assign({ label: inputName }, props); // props found, easy.

  var inputNameTree = hash_serializer({}, inputName, null);
  var attrName = Object.keys(inputNameTree)[0]; // inputName describes only single path

  props = inputProps[attrName];

  if (!isNaN(parseInt(attrName, 10))) {
    // skip number
    props = inputProps;
  } else if (props && props.type === 'nested') {
    // continue with nested props
    props = props.fields;
  } else if (props) {
    return props;
  } else {
    return undefined;
  }

  var newInputName = inputName.replace(attrName, '');
  var nextAttrName = Object.keys(inputNameTree[attrName])[0];

  // remove bracket
  newInputName = newInputName.replace('[' + nextAttrName + ']', '');
  newInputName = nextAttrName + newInputName;

  return inputPropsLookup(props, newInputName);
};

export const inputValueLookup = (serializedValues, inputName) => {
  if (!serializedValues || Object.keys(serializedValues).length === 0)
    return undefined;

  var value = serializedValues[inputName];
  if (typeof value !== 'undefined' && value !== null) return value; // value found, easy.

  // try to serialize and traverse
  var inputNameTree = hash_serializer({}, inputName, null);
  var attrName = Object.keys(inputNameTree)[0]; // inputName describes only single path

  var next = inputNameTree[attrName];
  if (next == null) return undefined; // if it's already leaf, then no chance to be found.

  var nextAttrName = Object.keys(next)[0];
  var newInputName = inputName.replace(attrName, '');

  // remove bracket
  newInputName = newInputName.replace('[' + nextAttrName + ']', '');

  if (newInputName === '') {
    return serializedValues[attrName]
      ? serializedValues[attrName][nextAttrName]
      : undefined;
  }
  newInputName = nextAttrName + newInputName;

  var nestedValues = serializedValues[attrName];
  return inputValueLookup(nestedValues, newInputName);
};

export const formSerialize = (form, options) => {
  options = options || { hash: true };
  return serialize(form, options);
};
