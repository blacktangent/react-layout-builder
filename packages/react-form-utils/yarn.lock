# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@alvinsj/form-serialize@^0.7.1":
  version "0.7.1"
  resolved "https://registry.yarnpkg.com/@alvinsj/form-serialize/-/form-serialize-0.7.1.tgz#73280c4c4b6971ba17c0913854747d59e5ccb165"

object-assign@^4.x.x:
  version "4.1.1"
  resolved "https://registry.yarnpkg.com/object-assign/-/object-assign-4.1.1.tgz#2109adc7965887cfc05cbbd442cac8bfbb360863"
