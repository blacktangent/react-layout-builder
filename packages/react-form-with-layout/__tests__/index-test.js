jest.unmock('react-form-utils');
jest.unmock('react-layout-builder');
jest.unmock('../src');

import React from 'react';
import PropTypes from 'prop-types';
import FormWithLayout from '../src';
import { inputPropsLookup, formSerialize } from 'react-form-utils';

import { mount } from 'enzyme';

const FIELDS = {
  age: {
    type: 'text'
  },
  name: {
    type: 'text',
    label: 'Full name'
  },
  email: {
    type: 'email',
    label: 'Email add.'
  },
  address: {
    type: 'text'
  },
  role: {
    type: 'hidden'
  }
};

const FormInput = props => {
  return (
    <input
      className={props.className}
      name={props.name}
      type={props.type}
      defaultValue={props.defaultValue}
    />
  );
};
FormInput.propTypes = {
  name: PropTypes.string,
  type: PropTypes.string,
  className: PropTypes.string,
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

describe('FormWithLayout', () => {
  var componentProps, createActions, component;

  beforeEach(() => {
    createActions = jest.fn();

    class FormOwner extends React.Component {
      constructor(props) {
        super(props);
        this._handleSubmit = this._handleSubmit.bind(this);
        this.renderButtons = this.renderButtons.bind(this);
      }

      render() {
        return (
          <FormWithLayout
            {...this.props}
            onSubmit={this._handleSubmit}
            getFieldProps={this.getFieldProps}
            renderLayout={this.getShortLayout}
            renderExpandedLayout={this.getFullLayout}
            renderField={(name, fieldProps) => <FormInput {...fieldProps} />}
            renderButtons={this.renderButtons}
          />
        );
      }

      getFieldProps(name) {
        return inputPropsLookup(FIELDS, name);
      }

      getFullLayout(builder, props) {
        const { layout, section, col } = builder;

        return layout(
          section(
            'main',
            [col(3, 'name', 'age'), col(6, 'email')],
            [col(6, 'address[0]', 'address[1]')]
          ),
          <input
            type="hidden"
            name={'role'}
            value={'husband'}
            key={'role-husband'}
          />,
          section(
            'main',
            [col(3, 'name', 'age'), col(6, 'email')],
            [col(6, 'address[0]', 'address[1]')]
          ),
          <input type="hidden" name={'role'} value={'wife'} key={'role-wife'} />
        );
      }

      getShortLayout(builder, props) {
        const { layout, section, col } = builder;

        return layout(
          section('husband', [col(6, 'name', 'email')]),
          <input
            type={'hidden'}
            name={'role'}
            value={'husband'}
            key={'role-husband'}
          />
        );
      }

      renderButtons(props) {
        return <button>Create</button>;
      }

      _handleSubmit(e) {
        const form = e.target;

        if (e) e.preventDefault();
        const values = formSerialize(form);
        createActions(values);
        this.setState({ defaultValues: values });

        if (this.props.onSubmit) this.props.onSubmit(e);
      }
    }
    FormOwner.propTypes = {
      onSubmit: PropTypes.func
    };

    componentProps = { showAll: true, defaultValues: { age: 0 } };
    component = mount(<FormOwner {...componentProps} />);
  });

  describe('#create', () => {
    it('renders form', () => {
      expect(component.find('form').length).toBe(1);
    });

    it('renders two same fields', () => {
      const form = component.find('form');
      expect(form.find('[name="name"]').length).toBe(2);
    });

    it('renders same number of fields when updated', () => {
      const form = component.find('form');
      expect(form.find('.field').length).toBe(10);

      component.update();
      expect(form.find('.field').length).toBe(10);
    });

    it('renders 1 layout and 2 sections', () => {
      expect(component.find('.layout').length).toBe(1);
      expect(component.find('.section').length).toBe(2);
    });

    describe('renders inputs', () => {
      var name, type, value;
      beforeAll(() => {
        var search = component.find(FormInput);
        expect(search.length).toBe(10);

        search = component.find('input');
        expect(search.length).toBe(12);

        name = n => {
          return search.at(n).node.getAttribute('name');
        };
        type = n => {
          return search.at(n).node.getAttribute('type');
        };
        value = n => {
          return search.at(n).node.value;
        };
      });

      it('renders all inputs', () => {
        expect(name(0)).toBe('name');
        expect(name(1)).toBe('age');
        expect(name(2)).toBe('email');
        expect(name(3)).toBe('address[0]');
        expect(name(4)).toBe('address[1]');
        expect(name(5)).toBe('role');
        expect(name(6)).toBe('name');
        expect(name(7)).toBe('age');
        expect(name(8)).toBe('email');
        expect(name(9)).toBe('address[0]');
        expect(name(10)).toBe('address[1]');
        expect(name(11)).toBe('role');
        expect(value(11)).toBe('wife');
      });

      it('renders input with correct type', () => {
        expect(name(2)).toBe('email');
        expect(type(2)).toBe('email');
      });

      it('renders nodes with value', () => {
        expect(name(5)).toBe('role');
        expect(value(5)).toBe('husband');

        expect(name(11)).toBe('role');
        expect(value(11)).toBe('wife');
      });

      it('renders input with 0 value', () => {
        expect(name(7)).toBe('age');
        expect(value(7)).toBe('0');
      });
    });

    it('renders lesser inputs when showAll=false', () => {
      component.setProps({ showAll: false });

      expect(component.props().showAll).toBe(false);
      var search = component.find(FormInput);
      expect(search.length).toBe(2);

      search = component.find('input');
      expect(search.length).toBe(3);
    });

    it('submit the form', () => {
      component.find('form').simulate('submit');
      expect(createActions).toBeCalled();
    });

    it('submit the form with values', () => {
      const name = 'garfield';
      component.setProps({ showAll: false, defaultValues: { name: name } });

      component.find('form').simulate('submit');
      expect(createActions.mock.calls[0][0]).toEqual({
        name: name,
        role: 'husband'
      });
    });

    it('passes the form node in onSubmit', () => {
      const grabContext = jest.fn();
      const onSubmit = jest.fn(function(e) {
        grabContext(e.target);
      });
      component.setProps({ onSubmit });

      const form = component.find('form');
      expect(form.length).toBe(1);
      form.simulate('submit');

      expect(onSubmit).toBeCalled();
      expect(grabContext).toBeCalledWith(form.node);
    });
  });
});
