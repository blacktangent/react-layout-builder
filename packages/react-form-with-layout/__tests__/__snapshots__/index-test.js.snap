// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormWithLayout #create renders inputs renders all inputs 1`] = `
Array [
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Full name"
    name="name"
    type="text"
>
    <input
        name="name"
        onChange={[Function]}
        type="text"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="age"
    name="age"
    type="text"
    value={0}
>
    <input
        name="age"
        onChange={[Function]}
        type="text"
        value={0}
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Email add."
    name="email"
    type="email"
>
    <input
        name="email"
        onChange={[Function]}
        type="email"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Address[0]"
    name="address[0]"
    type="text"
>
    <input
        name="address[0]"
        onChange={[Function]}
        type="text"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Address[1]"
    name="address[1]"
    type="text"
>
    <input
        name="address[1]"
        onChange={[Function]}
        type="text"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Full name"
    name="name"
    type="text"
>
    <input
        name="name"
        onChange={[Function]}
        type="text"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="age"
    name="age"
    type="text"
    value={0}
>
    <input
        name="age"
        onChange={[Function]}
        type="text"
        value={0}
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Email add."
    name="email"
    type="email"
>
    <input
        name="email"
        onChange={[Function]}
        type="email"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Address[0]"
    name="address[0]"
    type="text"
>
    <input
        name="address[0]"
        onChange={[Function]}
        type="text"
    />
</FormInput>,
  <FormInput
    formProps={
        Object {
            "getFieldProps": [Function],
            "onSubmit": [Function],
            "renderButtons": [Function],
            "renderExpandedLayout": [Function],
            "renderField": [Function],
            "renderLayout": [Function],
            "showAll": true,
            "values": Object {
              "age": 0,
            },
          }
    }
    label="Address[1]"
    name="address[1]"
    type="text"
>
    <input
        name="address[1]"
        onChange={[Function]}
        type="text"
    />
</FormInput>,
]
`;
